pkg/dockerregistry/server/prune/prune.go

```diff
@@ -19,6 +19,112 @@ import (
 	imageapiv1 "github.com/openshift/origin/pkg/image/apis/image/v1"
 )
 
+// Pruner defines a common set of operations for pruning
+type Pruner interface {
+	DeleteRepository(ctx context.Context, reponame string) error
+	DeleteManifestLink(ctx context.Context, svc distribution.ManifestService, reponame string, dgst digest.Digest) error
+	DeleteBlob(ctx context.Context, dgst digest.Digest) error
+}
+
+// DryRunPruner prints information about each object that going to remove.
+type DryRunPruner struct{}
+
+var _ Pruner = &DryRunPruner{}
+
+func (p *DryRunPruner) DeleteRepository(ctx context.Context, reponame string) error {
+	logger := context.GetLogger(ctx)
+	logger.Printf("Would delete repository: %s", reponame)
+	return nil
+}
+
+func (p *DryRunPruner) DeleteManifestLink(ctx context.Context, svc distribution.ManifestService, reponame string, dgst digest.Digest) error {
+	logger := context.GetLogger(ctx)
+	logger.Printf("Would delete manifest link: %s@%s", reponame, dgst)
+	return nil
+}
+
+func (p *DryRunPruner) DeleteBlob(ctx context.Context, dgst digest.Digest) error {
+	logger := context.GetLogger(ctx)
+	logger.Printf("Would delete blob: %s", dgst)
+	return nil
+}
+
+// RegistryPruner deletes objects.
+type RegistryPruner struct {
+	StorageDriver driver.StorageDriver
+}
+
+var _ Pruner = &RegistryPruner{}
+
+func (p *RegistryPruner) DeleteRepository(ctx context.Context, reponame string) error {
+	logger := context.GetLogger(ctx)
+	vacuum := storage.NewVacuum(ctx, p.StorageDriver)
+
+	logger.Debugln("Removing %s repository", reponame)
+	if err := vacuum.RemoveRepository(reponame); err != nil {
+		return fmt.Errorf("unable to remove the repository %s: %v", reponame, err)
+	}
+
+	return nil
+}
+
+func (p *RegistryPruner) DeleteManifestLink(ctx context.Context, svc distribution.ManifestService, reponame string, dgst digest.Digest) error {
+	logger := context.GetLogger(ctx)
+
+	logger.Printf("Deleting manifest link: %s@%s", reponame, dgst)
+	if err := svc.Delete(ctx, dgst); err != nil {
+		return fmt.Errorf("failed to delete the manifest link %s@%s: %v", reponame, dgst, err)
+	}
+
+	return nil
+}
+
+func (p *RegistryPruner) DeleteBlob(ctx context.Context, dgst digest.Digest) error {
+	vacuum := storage.NewVacuum(ctx, p.StorageDriver)
+
+	if err := vacuum.RemoveBlob(string(dgst)); err != nil {
+		return fmt.Errorf("failed to delete the blob %s: %v", dgst, err)
+	}
+
+	return nil
+}
+
+type garbageColletor struct {
+	Pruner Pruner
+
+	repoName string
+
+	manifestService distribution.ManifestService
+	manifestRepo    string
+	manifestLink    digest.Digest
+}
+
+func (gc *garbageColletor) SetRepository(repoName string) {
+	gc.repoName = repoName
+}
+
+func (gc *garbageColletor) SetManifestLink(svc distribution.ManifestService, repoName string, dgst digest.Digest) {
```

Comment:
*nit*: I'd prefer *add* instead of *set* to make it more aligned with real-world garbage collectors.

By marking an object for garbage collection, you don't want to clear the mark from any other marked object.

Reply:
i think Set is appropriate here because it's not appending the value to anything, it's replacing it.

Reply:
> it's replacing it.

And that's the problem. Those attributes holding objects marked for deletion should be arrays or set.

Requiring a user to call `Collect()` each time when he marks a single object for deletion (otherwise the garbage collector forgets earlier object) suggests broken design and makes the gb unusable for generic use.

But as I said, it's a nit since the gb is not exported.